imageJPG.ml

```diff
@@ -17,71 +17,308 @@
+
+    if first_dct_component <> 0 then
+      raise (Not_yet_implemented "Baseline decoding requires the first DCT component to be zero.");
+
+    let last_dct_component = get_bytes ich 1 |> int_of_str1 in
+
+    if last_dct_component <> 63 then
+      raise (Not_yet_implemented "Baseline decoding requires the last DCT component to be 63");
+
+    let approx = get_bytes ich 1 |> int_of_str1 in
+    let approx_high = (approx land 0xF0) lsr 4 in
+
+    if approx_high <> 0 then
+      raise (Not_yet_implemented "Baseline decoding requires the previous point transform to be zero");
+
+    let approx_low = approx land 0x0F in
+
+    if approx_low <> 0 then
+      raise (Not_yet_implemented "Baseline decoding requires the current point transform to be zero");
+
+    if !debug then
+      Printf.printf "SOS: length %d, component spec count: %d, first DCT component: %d, last DCT component: %d, approx high: %d, approx low: %d\n%!"
+        length compspeccount first_dct_component last_dct_component approx_high approx_low;
+    { component_specs; first_dct_component; last_dct_component; approx_high; approx_low }
+
+  (* Based on https://stackoverflow.com/a/48488655 *)
   let size ich =
-    let headers = read_header_data ich in
-    close_chunk_reader ich;
-
-    let rec find_SOF ls =
-      match ls with
-      | []        -> raise (Corrupted_image "No SOFn marker...");
-      | (m,d)::bs -> if m >= 0xc0 && m <= 0xcf && m <> 0xc4 && m <> 0xcc
-                     then (m - 0xc0, d)
-                     else find_SOF bs
+    let rec handle_marker ich c =
+      match string_of_marker c with
+      | "EOI" -> raise Not_found
+      | "SOF0" | "SOF1"  | "SOF2"  | "SOF3"  | "SOF5"  | "SOF6"  | "SOF7"
+      | "SOF9" | "SOF10" | "SOF11" | "SOF13" | "SOF14" | "SOF15" ->
+        let sof = parse_sof ich in
+        (sof.width, sof.height)
+      | marker ->
+        (if not (marker_stands_alone marker) then
+           let len = get_bytes ich 2 |> int_of_str2_be in
+           let data = get_bytes ich (len - 2) in
+           ignore data
+        );
+        find_next_marker ich handle_marker
     in
-    let (n, sof) = find_SOF headers in
-    (*
-    Printf.fprintf stderr "SOF%i: " n;
-    show_string_hex sof;
-    *)
-    let height = ((int_of_char sof.[1]) lsl 8) lor (int_of_char sof.[2]) in
-    let width = ((int_of_char sof.[3]) lsl 8) lor (int_of_char sof.[4]) in
-    width, height
+    try
+      find_next_marker ich handle_marker
+    with End_of_file ->
+      raise (Corrupted_image "Reached end of file while looking for SOF marker")
 
   let parsefile fn =
-    raise (Not_yet_implemented "ImageJPG.openfile") (* TODO  *)
+    let read_chunks = ref [] in
+
+    let sof = ref None in
+    let sos = ref None in
+    let image = ref None in
+
+    let rec handle_marker ich c =
+      let curr_ctype = string_of_marker c in
+      if !debug then
+        Printf.printf "Found marker %s\n%!" curr_ctype; (
```

Comment:
Can `curr_ctype` containing non-printable characters here? If so, consider using `%S` instead to escape it. Also, `eprintf`.

Reply:
>Can curr_ctype containing non-printable characters here?

No, `curr_ctype` is only the results of `string_to_marker`, which throws if it's an invalid marker.

Reply:
It seems to me that it's possible to have **valid** markers with non-printable characters, consider this: https://github.com/rlepigre/ocaml-imagelib/pull/14/files#diff-508dc458ff0e5a88c47970c398d99aa9R130